shared/src/token_info.rs

```diff
@@ -0,0 +1,81 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use contracts::ERC20;
+use ethcontract::{Http, Web3, H160};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+use mockall::*;
+
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Copy, Clone, Debug)]
+pub struct TokenInfo {
+    pub decimals: u8,
+}
+
+pub struct TokenInfoFetcher {
+    //pub factory: UniswapV2Factory,
+    pub web3: Web3<Http>,
+    //pub chain_id: u64,
+}
+
+#[automock]
+#[async_trait]
+pub trait TokenInfoFetching: Send + Sync {
+    /// Retrieves some token information from a token address.
+    async fn get_token_info(&self, address: &H160) -> Result<TokenInfo>;
+
+    /// Retrieves all token information.
+    /// Default implementation calls get_token_info for each token and ignores errors.
+    async fn get_token_infos(&self, addresses: &[H160]) -> Result<HashMap<H160, TokenInfo>> {
+        let mut result = HashMap::new();
```

Comment:
It would be more efficient to fetch all infos in one roundtrip to the node (by putting them in a single batch call like we do in the `PoolFetcher`), but given that we already cache responses I think it's not that important here.

On the other hand, since we don't really use the single get_token_info, we might as well only expose the plural one on the trait.

Reply:
Done!